File: frontend/lib/services/support_service.dart
import 'dart:convert';

import '../core/errors/app_exception.dart';
import 'api_service.dart';

class SupportService {
  Future<Map<String, dynamic>> createTicket({
    required String subject,
    required String description,
    String priority = 'medium',
    String? referenceId,
    String? referenceType,
  }) async {
    try {
      final body = <String, dynamic>{
        'subject': subject,
        'description': description,
        'priority': priority,
        if (referenceId != null) 'referenceId': referenceId,
        if (referenceType != null) 'referenceType': referenceType,
      };

      final response = await ApiService.post(
        '/support/tickets',
        body,
        requiresAuth: true,
      );
      return jsonDecode(response.body) as Map<String, dynamic>;
    } on AppException {
      rethrow;
    } catch (e) {
      throw NetworkException('Failed to create support ticket: ${e.toString()}');
    }
  }

  Future<List<dynamic>> getMyTickets() async {
    try {
      final response = await ApiService.get('/support/tickets', requiresAuth: true);
      final decoded = jsonDecode(response.body);
      if (decoded is List) return decoded;
      return const [];
    } on AppException {
      rethrow;
    } catch (e) {
      throw NetworkException('Failed to fetch support tickets: ${e.toString()}');
    }
  }

  Future<Map<String, dynamic>> getTicketById(String ticketId) async {
    try {
      final response = await ApiService.get('/support/tickets/$ticketId', requiresAuth: true);
      return jsonDecode(response.body) as Map<String, dynamic>;
    } on AppException {
      rethrow;
    } catch (e) {
      throw NetworkException('Failed to fetch support ticket: ${e.toString()}');
    }
  }

  Future<Map<String, dynamic>> submitFeedback({
    required String message,
    String type = 'app',
    int? rating,
    String? referenceId,
    String? referenceType,
  }) async {
    try {
      final body = <String, dynamic>{
        'message': message,
        'type': type,
        if (rating != null) 'rating': rating,
        if (referenceId != null) 'referenceId': referenceId,
        if (referenceType != null) 'referenceType': referenceType,
      };

      final response = await ApiService.post('/feedback', body, requiresAuth: true);
      return jsonDecode(response.body) as Map<String, dynamic>;
    } on AppException {
      rethrow;
    } catch (e) {
      throw NetworkException('Failed to submit feedback: ${e.toString()}');
    }
  }

  Future<List<dynamic>> getFaqs({String? category}) async {
    try {
      final endpoint = category == null ? '/faqs' : '/faqs?category=$category';
      final response = await ApiService.get(endpoint, requiresAuth: true);
      final decoded = jsonDecode(response.body);
      if (decoded is List) return decoded;
      return const [];
    } on AppException {
      rethrow;
    } catch (e) {
      throw NetworkException('Failed to fetch FAQs: ${e.toString()}');
    }
  }

  Future<Map<String, dynamic>> getSystemVersion() async {
    try {
      final response = await ApiService.get('/system/version');
      return jsonDecode(response.body) as Map<String, dynamic>;
    } on AppException {
      rethrow;
    } catch (e) {
      throw NetworkException('Failed to fetch system version: ${e.toString()}');
    }
  }
}

File: backend/src/support/support.service.ts
import { Injectable, NotFoundException } from '@nestjs/common';
import { InjectModel } from '@nestjs/mongoose';
import { Model, Types } from 'mongoose';
import {
  SupportTicket,
  SupportTicketDocument,
  TicketStatus,
  TicketPriority,
} from './schemas/support-ticket.schema';
import { Feedback, FeedbackDocument, FeedbackType } from './schemas/feedback.schema';
import { Faq, FaqDocument } from './schemas/faq.schema';

@Injectable()
export class SupportService {
  constructor(
    @InjectModel(SupportTicket.name)
    private ticketModel: Model<SupportTicketDocument>,
    @InjectModel(Feedback.name)
    private feedbackModel: Model<FeedbackDocument>,
    @InjectModel(Faq.name)
    private faqModel: Model<FaqDocument>,
  ) {}

  // ── Support Tickets ──────────────────────────────────────────────────────────

  async createTicket(
    userId: string,
    subject: string,
    description: string,
    priority: TicketPriority = TicketPriority.MEDIUM,
    referenceId?: string,
    referenceType?: string,
  ) {
    const ticket = new this.ticketModel({
      userId: new Types.ObjectId(userId),
      subject,
      description,
      priority,
      referenceId: referenceId ? new Types.ObjectId(referenceId) : undefined,
      referenceType,
    });
    return ticket.save();
  }

  async findMyTickets(userId: string) {
    return this.ticketModel
      .find({ userId: new Types.ObjectId(userId) })
      .sort({ updatedAt: -1 })
      .exec();
  }

  async findTicketById(ticketId: string, userId: string) {
    const ticket = await this.ticketModel
      .findOne({
        _id: new Types.ObjectId(ticketId),
        userId: new Types.ObjectId(userId),
      })
      .exec();

    if (!ticket) {
      throw new NotFoundException('Ticket not found');
    }
    return ticket;
  }

  async updateTicketStatus(
    ticketId: string,
    status: TicketStatus,
    adminId: string,
    resolution?: string,
  ) {
    const ticket = await this.ticketModel
      .findByIdAndUpdate(
        ticketId,
        {
          status,
          assignedTo: new Types.ObjectId(adminId),
          ...(resolution ? { resolution } : {}),
        },
        { new: true },
      )
      .exec();

    if (!ticket) throw new NotFoundException('Ticket not found');
    return ticket;
  }

  async findAllTickets(status?: TicketStatus, priority?: TicketPriority) {
    const filter: Record<string, unknown> = {};
    if (status) filter.status = status;
    if (priority) filter.priority = priority;
    return this.ticketModel
      .find(filter)
      .sort({ updatedAt: -1 })
      .populate('userId', 'name email')
      .exec();
  }

  // ── Feedback ────────────────────────────────────────────────────────────────

  async submitFeedback(
    userId: string,
    message: string,
    type: FeedbackType = FeedbackType.APP,
    rating?: number,
    referenceId?: string,
    referenceType?: string,
  ) {
    const feedback = new this.feedbackModel({
      userId: new Types.ObjectId(userId),
      message,
      type,
      rating,
      referenceId: referenceId ? new Types.ObjectId(referenceId) : undefined,
      referenceType,
    });
    return feedback.save();
  }

  // ── FAQ ─────────────────────────────────────────────────────────────────────

  async getFaqs(category?: string) {
    const filter: Record<string, unknown> = { isActive: true };
    if (category) filter.category = category;
    return this.faqModel.find(filter).sort({ sortOrder: 1, createdAt: 1 }).exec();
  }

  async createFaq(question: string, answer: string, category = 'general', sortOrder = 0) {
    const faq = new this.faqModel({ question, answer, category, sortOrder });
    return faq.save();
  }
}
